018_go_list_mod_graph_why/README.md

```diff
@@ -0,0 +1,459 @@
+
+We can answer that question using `go mod why` which shows us the shortest path from a package in the main module to any
+package in the `golang.org/x/tools` module:
+
+
+```
+{{PrintBlock "why tools" -}}
+```
+
+This seems to suggest that the dependency exists because of the requirements of
+`github.com/go-modules-by-example-forks/incomplete`. But the edge `github.com/go-modules-by-example-forks/incomplete ->
+golang.org/x/tools` is definitely not visible on our graph.
+
+Use `go list` again to find out more information about `golang.org/x/tools`:
+
+```
+{{PrintBlock "list tools" -}}
+```
+
+This shows that `golang.org/x/tools` is an indirect dependency.  Indirect requirements only arise when using modules
+that fail to state some of their own dependencies (this includes requirements that have not yet been converted to be
+modules) or when explicitly upgrading a module's dependencies ahead of its own stated requirements.
+
+We can also see this from the `// indirect` annotation in the main module's `go.mod`:
+
+```
+{{PrintBlockOut "go.mod" -}}
+```
+
+In this case, the indirect dependency comes about because the `go.mod` for
+`github.com/go-modules-by-example-forks/incomplete` fails to completely state its dependencies; `golang.org/x/tools` is
+missing.
+
+Returning to `go list`.
+
+`go list` gives information about packages. By default it lists the import path of packages matching the package
+patterns provided. With no arguments `go list` applies to the package in the current directory:
+
+```
+{{PrintBlock "go list" -}}
+```
+
+We can extend this to include the transitive dependencies of the named packages:
+
+```
+{{PrintBlock "go list -deps" | lineEllipsis 5 -}}
+```
+
+Use the `-f` flag to exclude standard library dependencies:
+
+
+```
+{{PrintBlock "go list nonstd" -}}
+```
+
+Looking at the original module requirement graph, we might ask ourselves, why are there no packages from the modules
+`github.com/davecgh/go-spew` and `github.com/kr/pty` in the `go list` output? Use `go mod why -m` to find the shortest
+path from a package in the main module to a package in each module.
+
+Firstly `github.com/davecgh/go-spew`:
+
+```
+{{PrintBlock "go mod why spew" -}}
+```
+
+We see that the package `github.com/davecgh/go-spew/spew` in the module `github.com/davecgh/go-spew` is one such path.
+We can therefore expect an identical answer if we ask the question of the package itself (no `-m` this time):
+
+```
+{{PrintBlock "go mod why spew pkg" -}}
+```
+
+Secondly, `github.com/kr/pty`:
+
+```
+{{PrintBlock "go mod why pty" -}}
+```
+
+This tells us that there is no package from the `github.com/kr/pty` module in the transitive import graph of of the main
+module (`{{PrintOut "main module"}}`).
+
+So why does the `github.com/kr/pty` appear in the module requirement graph? Currently, no command can give us
+the answer (something that is being tracked in [#27900](https://github.com/golang/go/issues/27900)).
```

Comment:
Perhaps mention that this issue will very likely be fixed in 1.12.

Reply:
Good point.